beacon_node/eth2_libp2p/src/peer_manager/peerdb.rs

```diff
@@ -515,10 +515,15 @@ impl<TSpec: EthSpec> PeerDB<TSpec> {
     /// Notifies the peer manager that the peer is undergoing a normal disconnect. Optionally tag
     /// the peer to be banned after the disconnect.
     pub fn notify_disconnecting(&mut self, peer_id: PeerId, to_ban_afterwards: bool) {
-        self.peers
-            .entry(peer_id)
-            .or_default()
-            .disconnecting(to_ban_afterwards);
+        let peer_info = self.peers.entry(peer_id).or_default();
+
+        if matches!(
+            peer_info.connection_status(),
+            PeerConnectionStatus::Disconnected { .. }
+        ) {
+            self.disconnected_peers = self.disconnected_peers.saturating_sub(1);
+        }
```

Comment:
Checking that I understand the logic here correctly:

We're about to change the peer's status from `Disconnected -> Disconnecting`, so we decrement the number of _disconnected_ peers. Usually a short time later we will re-increment it when the peer transitions from `Disconnecting -> Disconnected`, right?

Reply:
Yep, see, confuuuuusing.

Reply:
We shouldn't really be disconnecting, peers that are already disconnected, but it can happen that an old peer in our db tries to connect to us, and we have too many peers, so we dont register it as connected and just move it straight to disconnecting.

Reply:
Damn, yeah, that's complicated.